File: 20-actors/mizuho/methods/water_supply.py
"""water_supply — mizuho potable-water control loop (R0 :representative).

The runnable, tested core behind the `water_supply` cell. It proves a
community-scale supply actually holds pressure: a demand step (households open
taps) drops the reservoir level, the pump's secondary-PI loop drives inflow until
the level error integrates back to the service setpoint, and the modeled supply
restores service pressure.

This is the floating-point twin of an open-ot field-tier level/pressure loop.
mizuho constitutional gates apply: community-scale only (G3 — service population
is hard-capped here; a large municipal utility is N1, structurally unrepresentable),
no commercial water-utility software (G4 — this is a plain PI over a lumped tank,
not Veolia/Suez/Trojan firmware), Murakumo-only inference (G7 — not used in this
deterministic loop), and live actuation is consent-gated (G10 — this module is
offline sim only; cell.py .solve() stays Council-gated).
"""

from __future__ import annotations

from dataclasses import dataclass

from _substrate import PID, assert_civilian, simulate

# mizuho civilian-use allowlist (closed-world, N1). Water is for people + crops,
# never force. "supply" (potable distribution), "treat" (disinfection/filtration),
# "sample" (quality testing), "recycle" (greywater closed-loop), "irrigate"
# (mitsuho agricultural-grade dispatch).
PERMITTED_USES = ("supply", "treat", "sample", "recycle", "irrigate")

# G3 community-scale invariant: per-source service population is hard-capped.
# A request above the cap is N1 (a municipal utility) and is refused structurally.
MAX_SERVICE_POPULATION = 2500


class ReservoirPlant:
    """Community service-reservoir level dynamics (a Plant: measure/step).

    State is stored volume (litres). The pump command is inflow (L/s); a constant
    (settable) demand drains the tank. The controlled process variable is the
    water level (m) = volume / footprint-area; service pressure is proportional to
    head, so holding level holds pressure.

        dV/dt = inflow(command, L/s) - demand(L/s) - leak(level)
        level = V / area_m2 / 1000   (1 m^3 = 1000 L spread over area_m2)

    The reservoir is gravity-fed: distribution outflow ("leak") rises with head
    (`leak_coeff` L/s per metre of level), so the tank is self-regulating — a
    real first-order lag, not a pure integrator. Only a controller with integral
    action drives the level error to zero against a sustained demand, which is
    what the acceptance test asserts.
    """

    def __init__(
        self,
        area_m2: float = 20.0,
        level_m: float = 3.0,
        demand_lps: float = 0.0,
        max_level_m: float = 6.0,
        leak_coeff_lps_per_m: float = 100.0,
    ) -> None:
        self.area_m2 = area_m2
        self.max_level_m = max_level_m
        self._volume_l = level_m * area_m2 * 1000.0
        self._demand_lps = demand_lps
        self._leak_coeff = leak_coeff_lps_per_m

    def set_demand(self, demand_lps: float) -> None:
        """Apply a demand step (the disturbance the pump loop must reject)."""
        self._demand_lps = demand_lps

    @property
    def pressure_bar(self) -> float:
        """Service pressure (bar) ∝ static head (1 m water ≈ 0.0981 bar)."""
        return self.measure() * 0.0981

    def measure(self) -> float:
        return self._volume_l / (self.area_m2 * 1000.0)

    def step(self, command: float, dt: float) -> None:
        # command = pump inflow setpoint (L/s); cannot push the tank past its rim.
        # Gravity-fed distribution leak rises with head -> self-regulating tank.
        leak_lps = self._leak_coeff * self.measure()
        self._volume_l += (command - self._demand_lps - leak_lps) * dt
        if self._volume_l < 0.0:
            self._volume_l = 0.0
        max_v = self.max_level_m * self.area_m2 * 1000.0
        if self._volume_l > max_v:
            self._volume_l = max_v


@dataclass(frozen=True)
class WaterSupplyResult:
    """Outcome of a water-supply acceptance test (demand-step pressure recovery)."""

    use: str
    demand_step_lps: float
    setpoint_level_m: float
    final_level_m: float
    final_pressure_bar: float
    level_restored: bool
    settling_seconds: float
    service_population: int
    representative: bool  # G10: sims-only at R0


def commission_water_supply(
    demand_step_lps: float,
    use: str = "supply",
    setpoint_level_m: float = 3.0,
    area_m2: float = 20.0,
    service_population: int = 200,
    kp: float = 10.0,
    ki: float = 2.0,
    max_inflow_lps: float = 2000.0,
    steps: int = 4000,
    dt: float = 1.0,
) -> WaterSupplyResult:
    """Run the supply acceptance test. Raises (assert_civilian + G3) before any run.

    Apply `demand_step_lps`, run a secondary-PI pump loop, and confirm the level
    returns to `setpoint_level_m` (so service pressure is restored). Refuses a
    non-civilian use (N1) and refuses any request above the community-scale
    service-population cap (G3) — those are structurally unrepresentable here.
    """
    assert_civilian(use, PERMITTED_USES)  # N1 gate before any actuation modelling
    if service_population > MAX_SERVICE_POPULATION:
        from _substrate import SafetyError

        raise SafetyError(
            f"G3: service_population {service_population} exceeds the community-scale "
            f"cap {MAX_SERVICE_POPULATION}; a larger system is N1 (a municipal utility) "
            "and is structurally unrepresentable in mizuho"
        )

    tank = ReservoirPlant(area_m2=area_m2, level_m=setpoint_level_m, demand_lps=0.0)
    tank.set_demand(demand_step_lps)
    # Pump inflow is non-negative (a community pump cannot suck the tank down).
    pid = PID(kp=kp, ki=ki, out_min=0.0, out_max=max_inflow_lps)
    res = simulate(tank, pid, setpoint=setpoint_level_m, steps=steps, dt=dt, tol=1e-3)

    settling_seconds = res.settling_step * dt if res.settling_step >= 0 else -1.0
    return WaterSupplyResult(
        use=use,
        demand_step_lps=demand_step_lps,
        setpoint_level_m=setpoint_level_m,
        final_level_m=round(res.final_value, 4),
        final_pressure_bar=round(tank.pressure_bar, 4),
        level_restored=res.converged,
        settling_seconds=round(settling_seconds, 3),
        service_population=service_population,
        representative=True,
    )


def to_datoms(result: WaterSupplyResult, source_id: str) -> dict:
    """Project a supply acceptance result into kotoba EAVT-shaped datoms (G6/G9).

    Aggregate-only (no per-household consumption PII). The transactor appends these
    to the canonical Datom log; here we return the entity map a transactor writes.
    """
    return {
        ":water.supply/source-id": source_id,
        ":water.supply/use": result.use,
        ":water.supply/demand-step-lps": result.demand_step_lps,
        ":water.supply/setpoint-level-m": result.setpoint_level_m,
        ":water.supply/final-level-m": result.final_level_m,
        ":water.supply/final-pressure-bar": result.final_pressure_bar,
        ":water.supply/level-restored": result.level_restored,
        ":water.supply/settling-seconds": result.settling_seconds,
        ":water.supply/service-population": result.service_population,  # aggregate, ≤ G3 cap
        ":water.supply/representative": result.representative,          # G10
        ":water.supply/server-held-key": False,                        # no-server-key
        ":water.supply/dry-run": True,                                 # G10: R0 offline only
    }

File: 20-actors/mizuho/methods/water_supply.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/water_supply.py (reuses the clj control substrate).
(ns mizuho.methods.water-supply
  "water_supply — mizuho potable-water control loop (R0 :representative).

  Proves a community-scale supply holds pressure: a demand step (households open taps) drops
  the reservoir level, the pump's secondary-PI loop drives inflow until the level error
  integrates back to the service setpoint, and the modeled supply restores service pressure.

  mizuho gates: G3 community-scale only (service population hard-capped; a municipal utility is
  N1, structurally unrepresentable) · G4 plain PI over a lumped tank (never commercial
  water-utility firmware) · N1 civilian-use gate (assert-civilian) · G10 live actuation
  consent-gated; this module is offline sim only.

  Run:  bb --classpath 20-actors 20-actors/mizuho/methods/water_supply.clj"
  (:require [mizuho.methods.substrate :as s]))

;; civilian-use allowlist (closed-world, N1): water is for people + crops, never force.
(def permitted-uses #{"supply" "treat" "sample" "recycle" "irrigate"})

;; G3 community-scale invariant: per-source service population is hard-capped. Above the cap is
;; N1 (a municipal utility) and is refused structurally.
(def MAX-SERVICE-POPULATION 2500)

(defn reservoir-plant
  "Community service-reservoir level dynamics (a Plant: measure/step). State = stored volume
  (L); pump command = inflow (L/s); a constant demand drains it; gravity-fed distribution leak
  rises with head (self-regulating, real first-order lag). Level (m) = V / area / 1000."
  [{:keys [area-m2 level-m demand-lps max-level-m leak-coeff-lps-per-m]
    :or {area-m2 20.0 level-m 3.0 demand-lps 0.0 max-level-m 6.0 leak-coeff-lps-per-m 100.0}}]
  (let [vol (atom (* level-m area-m2 1000.0))
        demand (atom (double demand-lps))
        measure (fn [] (/ @vol (* area-m2 1000.0)))]
    {:measure measure
     :area-m2 area-m2
     :set-demand! (fn [d] (reset! demand (double d)))
     :pressure-bar (fn [] (* (measure) 0.0981))   ; 1 m water ≈ 0.0981 bar
     :step (fn [command dt]
             (let [leak (* leak-coeff-lps-per-m (measure))]
               (swap! vol #(+ % (* (- command @demand leak) dt)))
               (when (< @vol 0.0) (reset! vol 0.0))
               (let [max-v (* max-level-m area-m2 1000.0)]
                 (when (> @vol max-v) (reset! vol max-v)))))}))

(defn commission-water-supply
  "Run the supply acceptance test. Raises (assert-civilian + G3) before any run. Apply
  `demand-step-lps`, run a secondary-PI pump loop, and confirm the level returns to
  `setpoint-level-m` (service pressure restored)."
  [{:keys [demand-step-lps use setpoint-level-m area-m2 service-population kp ki max-inflow-lps steps dt]
    :or {use "supply" setpoint-level-m 3.0 area-m2 20.0 service-population 200
         kp 10.0 ki 2.0 max-inflow-lps 2000.0 steps 4000 dt 1.0}}]
  (s/assert-civilian use permitted-uses)               ; N1 gate before any actuation modelling
  (when (> service-population MAX-SERVICE-POPULATION)
    (s/safety-error (str "G3: service-population " service-population " exceeds the community-scale "
                         "cap " MAX-SERVICE-POPULATION "; a larger system is N1 (a municipal utility) "
                         "and is structurally unrepresentable in mizuho")))
  (let [tank (reservoir-plant {:area-m2 area-m2 :level-m setpoint-level-m :demand-lps 0.0})
        _ ((:set-demand! tank) demand-step-lps)
        pid (s/make-pid {:kp kp :ki ki :out-min 0.0 :out-max max-inflow-lps})
        controller {:reset (fn [] (s/pid-reset! pid)) :step (fn [e dt] (s/pid-step! pid e dt))}
        res (s/simulate tank controller setpoint-level-m steps dt :tol 1e-3)
        settling-seconds (if (>= (:settling-step res) 0) (* (:settling-step res) dt) -1.0)]
    {:use use
     :demand-step-lps demand-step-lps
     :setpoint-level-m setpoint-level-m
     :final-level-m (/ (Math/round (* (:final-value res) 10000.0)) 10000.0)
     :final-pressure-bar (/ (Math/round (* ((:pressure-bar tank)) 10000.0)) 10000.0)
     :level-restored (:converged res)
     :settling-seconds (/ (Math/round (* settling-seconds 1000.0)) 1000.0)
     :service-population service-population
     :representative true}))

(defn to-datoms
  "Project a supply acceptance result into kotoba EAVT-shaped datoms (aggregate-only; no
  per-household consumption PII)."
  [result source-id]
  {:water.supply/source-id source-id
   :water.supply/use (:use result)
   :water.supply/demand-step-lps (:demand-step-lps result)
   :water.supply/setpoint-level-m (:setpoint-level-m result)
   :water.supply/final-level-m (:final-level-m result)
   :water.supply/final-pressure-bar (:final-pressure-bar result)
   :water.supply/level-restored (:level-restored result)
   :water.supply/settling-seconds (:settling-seconds result)
   :water.supply/service-population (:service-population result)  ; aggregate, ≤ G3 cap
   :water.supply/representative (:representative result)          ; G10
   :water.supply/server-held-key false                           ; no-server-key
   :water.supply/dry-run true})                                  ; G10: R0 offline only

(defn main [& _]
  (let [r (commission-water-supply {:demand-step-lps 20.0})]
    (println (format "mizuho water-supply: use=%s demand=%.1f L/s final-level=%.4f m pressure=%.4f bar restored=%s settle=%.0fs pop=%d"
                     (:use r) (double (:demand-step-lps r)) (double (:final-level-m r))
                     (double (:final-pressure-bar r)) (:level-restored r)
                     (double (:settling-seconds r)) (:service-population r)))))

(when (= *file* (System/getProperty "babashka.file"))
  (main))
